src/ert/com_protocol/_state_machine.py

```diff
@@ -0,0 +1,240 @@
+# pylint: disable=E1101 # might be fixed when we migrate to proto3
+import logging
+from functools import singledispatchmethod
+from typing import Optional, Union
+
+import ert.com_protocol._schema_pb2 as pb2
+
+logger = logging.getLogger(__name__)
+
+
+class _StateHandle:
+    """The :class:`_StateHandle` implements a mechanism for handling the protobuf
+    defined state machine.
+    General note:
+    We cannot directly assign values to a protobuf-map; eg:
+    experiment.ensembles[ens_id] = Ensemble(id=ens_id)
+    Useful links:
+    https://developers.google.com/protocol-buffers/docs/reference/csharp/class/google/protobuf/well-known-types/struct
+    https://developers.google.com/protocol-buffers/docs/reference/python-generated#map-fields
+    """
+
+    def __init__(self) -> None:
+        self._experiment: Optional[pb2.Experiment] = None
+
+    def get_experiment(self, _id: pb2.ExperimentId) -> pb2.Experiment:
+        logger.debug(f"experiment id: {_id}")
+        if self._experiment:
+            if self._experiment.id.id != _id.id:
+                raise ExperimentStateMachine.IllegalStateUpdate(
```

Comment:
Is "getting an experiment" a "state update"? (since an IllegalStateUpdate is raised)

Reply:
This is to verify that another (or late?) experiment is to update the current experiment. Currently this answers the experiment registry needs: https://github.com/equinor/ert/blob/3dcc8c29c173664dcf7949ddd25ab1e96be6a401/src/ert/experiment_server/_registry.py
We'll see if there is a need for it later on since currently there should be only one experiment running. The plan is to have the gui run different experiments every time we press `start simulation` button.